File: src/main/java/main/commands/slashcommands/commands/audio/FavouriteTracksCommand.java
package main.commands.slashcommands.commands.audio;

import main.audiohandlers.RobertifyAudioManager;
import main.commands.prefixcommands.CommandContext;
import main.commands.prefixcommands.ICommand;
import main.constants.RobertifyTheme;
import main.constants.Toggles;
import main.constants.TrackSource;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.component.interactions.selectionmenu.SelectMenuOption;
import main.utils.database.mongodb.cache.FavouriteTracksCache;
import main.utils.json.restrictedchannels.RestrictedChannelsConfig;
import main.utils.json.themes.ThemesConfig;
import main.utils.json.toggles.TogglesConfig;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocaleMessage;
import main.utils.pagination.Pages;
import net.dv8tion.jda.api.EmbedBuilder;
import net.dv8tion.jda.api.entities.*;
import net.dv8tion.jda.api.entities.channel.concrete.TextChannel;
import net.dv8tion.jda.api.events.interaction.component.SelectMenuInteractionEvent;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.interactions.commands.OptionType;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.jetbrains.annotations.NotNull;

import javax.script.ScriptException;
import java.util.ArrayList;
import java.util.List;

public class FavouriteTracksCommand extends AbstractSlashCommand implements ICommand {
    @Override
    public void handle(CommandContext ctx) throws ScriptException {
        final var args = ctx.getArgs();
        final var member=  ctx.getMember();
        final var msg = ctx.getMessage();
        final var channel = ctx.getChannel();
        final var localeManager = LocaleManager.getLocaleManager(msg.getGuild());

        if (args.isEmpty()) {
            handeList(channel, member);
        } else {
            switch (args.get(0).toLowerCase()) {
                case "add" -> msg.replyEmbeds(handleAdd(ctx.getGuild(), member)).queue();
                case "remove" -> {
                    final var guild = msg.getGuild();

                    if (args.size() < 2) {
                        msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, localeManager.getMessage(RobertifyLocaleMessage.FavouriteTracksMessages.FT_INVALID_ID)).build())
                                .queue();
                        return;
                    }

                    if (!GeneralUtils.stringIsInt(args.get(1))) {
                        msg.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, localeManager.getMessage(RobertifyLocaleMessage.FavouriteTracksMessages.FT_INVALID_SOURCE)).build())
                                .queue();
                        return;
                    }

                    final int id = Integer.parseInt(args.get(1));
                    msg.replyEmbeds(handleRemove(ctx.getGuild(), ctx.getAuthor(), id)).queue();
                }
                case "clear" -> msg.replyEmbeds(handleClear(ctx.getGuild(), ctx.getAuthor())).queue();
                case "list" -> handeList(channel, member);
            }
        }
    }

    @Override
    public boolean isPremiumCommand() {
        return true;
    }

    public MessageEmbed handleAdd(Guild guild, @NotNull Member member) {
        final var config = FavouriteTracksCache.getInstance();
        final var musicManager = RobertifyAudioManager.getInstance().getMusicManager(guild);
        final var audioPlayer = musicManager.getPlayer();
        final var playingTrack = audioPlayer.getPlayingTrack();
        final var memberVoiceState = member.getVoiceState();

        if (!memberVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL).build();

        final var selfVoiceState = guild.getSelfMember().getVoiceState();

        if (!selfVoiceState.inAudioChannel())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.MUST_BE_PLAYING).build();

        if (!selfVoiceState.getChannel().equals(memberVoiceState.getChannel()))
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention())).build();

        if (playingTrack == null)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.MUST_BE_PLAYING).build();

        final var trackInfo = playingTrack.getInfo();

        String id = trackInfo.identifier;
        TrackSource source;

        switch (playingTrack.getSourceManager().getSourceName()) {
            case "spotify" -> source = TrackSource.SPOTIFY;
            case "deezer" -> source = TrackSource.DEEZER;
            case "youtube" -> source = TrackSource.YOUTUBE;
            default -> {
                return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.FT_INVALID_SOURCE).build();
            }
        }

        try {
            config.addTrack(member.getIdLong(), id, trackInfo.title, trackInfo.author, source);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.FAV_TRACK_ADDED, Pair.of("{title}", playingTrack.getInfo().title), Pair.of("{author}", playingTrack.getInfo().author)).build();
        } catch (IllegalArgumentException e) {
            return RobertifyEmbedUtils.embedMessage(guild, e.getMessage()).build();
        } catch (Exception e) {
            e.printStackTrace();
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.UNEXPECTED_ERROR).build();
        }
    }

    private MessageEmbed handleRemove(Guild guild, User user, int id) {

        if (id <= 0)
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.ID_GT_ZERO).build();

        final var config = FavouriteTracksCache.getInstance();
        final var trackList = config.getTracks(user.getIdLong());

        if (id > trackList.size())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.ID_OUT_OF_BOUNDS).build();

        try {
            config.removeTrack(user.getIdLong(), id-1);
            final var trackRemoved = trackList.get(id-1);
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.FAV_TRACK_REMOVED, Pair.of("{title}", trackRemoved.title()), Pair.of("{author}", trackRemoved.author())).build();
        } catch (NullPointerException e) {
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.NO_FAV_TRACKS).build();
        } catch (Exception e) {
            e.printStackTrace();
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.UNEXPECTED_ERROR).build();
        }
    }

    private MessageEmbed handleClear(Guild guild, User user) {
        final var config = FavouriteTracksCache.getInstance();
        final var trackList = config.getTracks(user.getIdLong());

        if (trackList.isEmpty())
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.NO_FAV_TRACKS).build();

        try {
            config.clearTracks(user.getIdLong());
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.FAV_TRACKS_CLEARED).build();
        } catch (NullPointerException e) {
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.NO_FAV_TRACKS).build();
        } catch (Exception e) {
            e.printStackTrace();
            return RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.UNEXPECTED_ERROR).build();
        }
    }

    private void handeList(TextChannel channel, Member member) {
        final var config = FavouriteTracksCache.getInstance();
        final var guild = member.getGuild();

        final var tracks = config.getTracks(member.getIdLong());

        if (tracks.isEmpty()) {
            channel.sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.NO_FAV_TRACKS).build())
                    .queue();
            return;
        }

        final List<SelectMenuOption> list = new ArrayList<>();
        final var localeManager = LocaleManager.getLocaleManager(guild);

        for (final var track : tracks)
            list.add(
                    SelectMenuOption.of(
                            localeManager.getMessage(
                                    RobertifyLocaleMessage.FavouriteTracksMessages.FT_SELECT_MENU_OPTION,
                                    Pair.of("{title}", track.title().substring(0, Math.min(75, track.title().length()))),
                                    Pair.of("{author}", track.author().substring(0, Math.min(20, track.author().length())))
                            ),
                            "favouriteTrack:" + track.id() + ":" + track.source())
            );

        final var theme = new ThemesConfig(guild).getTheme();
        setDefaultEmbed(member, tracks, theme);
        Pages.paginateMenu(channel, member.getUser(),  list, 0,true);
    }

    private void handeSlashList(SlashCommandInteractionEvent event) {
        final var config = FavouriteTracksCache.getInstance();
        final var member = event.getMember();
        final var guild = member.getGuild();

        final var tracks = config.getTracks(member.getIdLong());

        if (tracks.isEmpty()) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.NO_FAV_TRACKS).build())
                    .queue();
            return;
        }

        final List<SelectMenuOption> list = new ArrayList<>();
        final var localeManager = LocaleManager.getLocaleManager(guild);

        for (final var track : tracks)
            list.add(
                    SelectMenuOption.of(
                            localeManager.getMessage(
                                    RobertifyLocaleMessage.FavouriteTracksMessages.FT_SELECT_MENU_OPTION,
                                    Pair.of("{title}", track.title().substring(0, Math.min(75, track.title().length()))),
                                    Pair.of("{author}", track.author().substring(0, Math.min(20, track.author().length())))
                            ),
                            "favouriteTrack:" + track.id() + ":" + track.source())
            );

        final var theme = new ThemesConfig(guild).getTheme();
        setDefaultEmbed(member, tracks, theme);
        Pages.paginateMenu(event, list, 0,true);
    }

    @Override
    public String getName() {
        return "favouritetracks";
    }

    @Override
    public String getHelp(String prefix) {
        return "Aliases: `"+GeneralUtils.listToString(getAliases())+"`\n\n" +
                "This command allows you to interact with tracks you may really like when" +
                " using the bot! Want to save some really good songs for later? No problem! " +
                "We'll store it for you.\n\n" + getUsages(prefix);
    }

    @Override
    public String getUsages(String prefix) {
        return "**__Usages__**\n" +
                "`" + prefix + "favouritetracks add` *(Add the current song as a favourite track)*\n" +
                "`" + prefix + "favouritetracks remove <id>` *(Remove a specified song as a favourite track)*\n" +
                "`" + prefix + "favouritetracks` *(View all your favourite tracks)*\n";
    }

    @Override
    public List<String> getAliases() {
        return List.of("favs", "fav", "favoritetracks", "favtracks");
    }

    @Override
    protected void buildCommand() {
        setCommand(
                getBuilder()
                        .setName("favouritetracks")
                        .setDescription("Interact with your favourite tracks using this command!")
                        .addSubCommands(
                                SubCommand.of(
                                        "view",
                                        "View all of your favourite tracks!"
                                ),
                                SubCommand.of(
                                        "add",
                                        "Add the current playing song as one of your favourites!"
                                ),
                                SubCommand.of(
                                        "remove",
                                        "Remove a specific track as a favourite track",
                                        List.of(
                                                CommandOption.of(
                                                        OptionType.INTEGER,
                                                        "id",
                                                        "The ID of the song to remove",
                                                        true
                                                )
                                        )
                                ),
                                SubCommand.of(
                                        "clear",
                                        "Clear all of your favourite tracks!"
                                )
                        )
                        .setPremium()
                        .build()
        );
    }

    @Override
    public String getHelp() {
        return "This command allows you to interact with tracks you may really like when" +
                " using the bot! Want to save some really good songs for later? No problem! " +
                "We'll store it for you.\n\n" + getUsages();
    }

    @Override
    public String getUsages() {
        return """
                **__Usages__**
                `/favouritetracks add` *(Add the current song as a favourite track)*
                `/favouritetracks remove <id>` *(Remove a specified song as a favourite track)*
                `/favouritetracks` *(View all your favourite tracks)*
                """;
    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        if (!checksWithPremium(event)) return;
        sendRandomMessage(event);

        switch (event.getSubcommandName()) {
            case "view" -> handeSlashList(event);
            case "add" -> event.replyEmbeds(handleAdd(event.getGuild(), event.getMember())).setEphemeral(true).queue();
            case "remove" -> {
                int id = (int)event.getOption("id").getAsLong();
                event.replyEmbeds(handleRemove(event.getGuild(), event.getUser(), id))
                        .setEphemeral(true)
                        .queue();
            }
            case "clear" -> event.replyEmbeds(handleClear(event.getGuild(), event.getUser())).setEphemeral(true).queue();
        }
    }

    @Override
    public void onSelectMenuInteraction(@NotNull SelectMenuInteractionEvent event) {
        if (!event.getSelectMenu().getId().startsWith("menupage")) return;
        if (!event.getUser().getId().equals(event.getComponentId().split(":")[1]))
            return;

        final var selectionOption = event.getSelectedOptions().get(0).getValue();

        if (!selectionOption.startsWith("favouriteTrack")) return;

        final String id = selectionOption.split(":")[1];
        final TrackSource source = TrackSource.parse(selectionOption.split(":")[2]);
        final var audioManager = RobertifyAudioManager.getInstance();
        final var guild = event.getGuild();
        final var memberVoiceState = event.getMember().getVoiceState();
        final var selfVoiceState = event.getGuild().getSelfMember().getVoiceState();

        if (!memberVoiceState.inAudioChannel()) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL).build())
                    .setEphemeral(true)
                    .queue();
            return;
        }

        if (selfVoiceState.inAudioChannel() && !memberVoiceState.getChannel().equals(selfVoiceState.getChannel())) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.SAME_VOICE_CHANNEL_LOC, Pair.of("{channel}", selfVoiceState.getChannel().getAsMention()))
                            .build())
                    .setEphemeral(true)
                    .queue();
            return;
        } else if (!selfVoiceState.inAudioChannel()) {
            if (new TogglesConfig(guild).getToggle(Toggles.RESTRICTED_VOICE_CHANNELS)) {
                final var restrictedChannelsConfig = new RestrictedChannelsConfig(guild);
                final var localeManager = LocaleManager.getLocaleManager(guild);
                if (!restrictedChannelsConfig.isRestrictedChannel(memberVoiceState.getChannel().getIdLong(), RestrictedChannelsConfig.ChannelType.VOICE_CHANNEL)) {
                    event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.CANT_JOIN_CHANNEL) +
                                    (!restrictedChannelsConfig.getRestrictedChannels(
                                            RestrictedChannelsConfig.ChannelType.VOICE_CHANNEL
                                    ).isEmpty()
                                            ?
                                            localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.RESTRICTED_TO_JOIN, Pair.of("{channels}", restrictedChannelsConfig.restrictedChannelsToString(
                                                    RestrictedChannelsConfig.ChannelType.VOICE_CHANNEL
                                            )))
                                            :
                                            localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.NO_VOICE_CHANNEL)
                                    )
                            ).build())
                            .setEphemeral(true)
                            .queue();
                    return;
                }
            }
        }

        event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.FT_ADDING_TO_QUEUE).build())
                .setEphemeral(true)
                .queue();

        event.getChannel().asTextChannel().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.FavouriteTracksMessages.FT_ADDING_TO_QUEUE_2).build()).queue(addingMsg -> {
            switch (source) {
                case DEEZER -> audioManager.loadAndPlay("https://www.deezer.com/us/track/" + id, selfVoiceState, memberVoiceState, event.getChannel().asTextChannel(), event.getUser(), addingMsg, false);
                case SPOTIFY -> audioManager.loadAndPlay("https://open.spotify.com/track/" + id, selfVoiceState, memberVoiceState, event.getChannel().asTextChannel(), event.getUser(), addingMsg, false);
                case YOUTUBE -> audioManager.loadAndPlay(id, selfVoiceState, memberVoiceState, event.getChannel().asTextChannel(), event.getUser(), addingMsg, false);
            }
        });
    }

    public static void setDefaultEmbed(Member member, List<FavouriteTracksCache.Track> tracks, RobertifyTheme theme) {
        final var localeManager  = LocaleManager.getLocaleManager(member.getGuild());
        Pages.setEmbedStyle(
                () -> new EmbedBuilder()
                        .setColor(theme.getColor())
                        .setTitle(localeManager.getMessage(RobertifyLocaleMessage.FavouriteTracksMessages.FT_EMBED_TITLE, Pair.of("{user}", member.getEffectiveName())))
                        .setFooter(localeManager.getMessage(RobertifyLocaleMessage.FavouriteTracksMessages.FT_EMBED_FOOTER, Pair.of("{tracks}", String.valueOf(tracks.size()))))
                        .setDescription(localeManager.getMessage(RobertifyLocaleMessage.FavouriteTracksMessages.FT_EMBED_DESCRIPTION))
                        .setThumbnail(member.getEffectiveAvatarUrl())
        );
    }
}

File: src/main/kotlin/main/commands/slashcommands/audio/FavouriteTracksCommand.kt
package main.commands.slashcommands.audio

import dev.minn.jda.ktx.util.SLF4J
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.utils.author
import main.audiohandlers.utils.source
import main.audiohandlers.utils.title
import main.commands.slashcommands.SlashCommandManager.getRequiredOption
import main.constants.RobertifyTheme
import main.constants.TrackSource
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.selectionmenu.StringSelectMenuOption
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.Command
import main.utils.component.interactions.slashcommand.models.CommandOption
import main.utils.component.interactions.slashcommand.models.SubCommand
import main.utils.database.mongodb.cache.FavouriteTracksCache
import main.utils.json.themes.ThemesConfig
import main.utils.locale.LocaleManager
import main.utils.locale.messages.FavouriteTracksMessages
import main.utils.locale.messages.GeneralMessages
import main.utils.pagination.PaginationHandler
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import net.dv8tion.jda.api.entities.MessageEmbed
import net.dv8tion.jda.api.entities.User
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.StringSelectInteractionEvent
import net.dv8tion.jda.api.interactions.commands.OptionType

class FavouriteTracksCommand : AbstractSlashCommand(
    Command(
        name = "favouritetracks",
        description = "Interact with your favourite tracks using this command.",
        subcommands = listOf(
            SubCommand(
                name = "view",
                description = "View all of your favourite tracks."
            ),
            SubCommand(
                name = "add",
                description = "Add the currently playing song as one of your favourites."
            ),
            SubCommand(
                name = "remove",
                description = "Remove a specific track as a favourite track.",
                listOf(
                    CommandOption(
                        type = OptionType.INTEGER,
                        name = "id",
                        description = "The ID of the song to remove."
                    )
                )
            ),
            SubCommand(
                name = "Clear",
                description = "Clear all of your favourite tracks"
            )
        ),
        isPremium = true
    )
) {

    companion object {
        private val logger by SLF4J
    }

    override val help: String
        get() = "This command allows you to interact with tracks you may really like when" +
                " using the bot! Want to save some really good songs for later? No problem! " +
                "We'll store it for you.\n\n" +
                """
                **__Usages__**
                `/favouritetracks add` *(Add the current song as a favourite track)*
                `/favouritetracks remove <id>` *(Remove a specified song as a favourite track)*
                `/favouritetracks` *(View all your favourite tracks)*
                """

    override suspend fun handle(event: SlashCommandInteractionEvent) {
        when (event.subcommandName) {
            "view" -> handleList(event)
            "add" -> event.replyEmbed { handleAdd(event.guild!!, event.member!!) }.setEphemeral(true).queue()
            "remove" -> {
                val id = event.getRequiredOption("id").asInt
                event.replyEmbed { handleRemove(event.guild!!, event.user, id) }
                    .setEphemeral(true)
                    .queue()
            }

            "clear" -> event.replyEmbed { handleClear(event.guild!!, event.user) }.setEphemeral(true).queue()
        }
    }

    private fun handleList(event: SlashCommandInteractionEvent) {
        val config = FavouriteTracksCache.instance
        val member = event.member!!
        val guild = member.guild
        val tracks = config.getTracks(member.idLong)

        if (tracks.isEmpty()) {
            event.replyEmbed {
                embed(FavouriteTracksMessages.NO_FAV_TRACKS)
            }
                .setEphemeral(true)
                .queue()

            return
        }

        val list = mutableListOf<StringSelectMenuOption>()
        val localeManager = LocaleManager[guild]

        tracks.forEach { track ->
            list.add(
                StringSelectMenuOption(
                    label = localeManager.getMessage(
                        FavouriteTracksMessages.FT_SELECT_MENU_OPTION,
                        Pair("{title}", track.title.substring(0, track.title.length.coerceAtMost(75))),
                        Pair("{author}", track.author.substring(0, track.author.length.coerceAtMost(20)))
                    ),
                    value = "favouriteTrack:${track.id}:${track.source}"
                )
            )
        }

        val theme = ThemesConfig(guild).theme
        setDefaultEmbed(member, tracks, theme)
        PaginationHandler.paginateMenu(event, list)
    }

    private fun handleClear(guild: Guild, user: User): MessageEmbed {
        val config = FavouriteTracksCache.instance
        val trackList = config.getTracks(user.idLong)

        if (trackList.isEmpty())
            return RobertifyEmbedUtils.embedMessage(
                guild,
                FavouriteTracksMessages.NO_FAV_TRACKS
            ).build()

        return try {
            config.clearTracks(user.idLong)
            RobertifyEmbedUtils.embedMessage(
                guild,
                FavouriteTracksMessages.FAV_TRACKS_CLEARED
            ).build()
        } catch (e: NullPointerException) {
            RobertifyEmbedUtils.embedMessage(guild, FavouriteTracksMessages.NO_FAV_TRACKS)
                .build()
        } catch (e: Exception) {
            logger.error("Unexpected error", e)
            RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.UNEXPECTED_ERROR).build()
        }
    }

    private fun handleRemove(guild: Guild, user: User, id: Int): MessageEmbed {
        if (id <= 0)
            return RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.ID_GT_ZERO)
                .build()

        val config = FavouriteTracksCache.instance
        val trackList = config.getTracks(user.idLong)

        if (id > trackList.size)
            return RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.ID_OUT_OF_BOUNDS)
                .build()

        return try {
            config.removeTrack(user.idLong, id - 1)
            val trackRemoved = trackList.get(id - 1)
            RobertifyEmbedUtils.embedMessage(
                guild,
                FavouriteTracksMessages.FAV_TRACK_REMOVED,
                Pair("{title}", trackRemoved.title),
                Pair("{author}", trackRemoved.author)
            ).build()
        } catch (e: NullPointerException) {
            RobertifyEmbedUtils.embedMessage(guild, FavouriteTracksMessages.NO_FAV_TRACKS)
                .build()
        } catch (e: Exception) {
            logger.error("Unexpected error", e)
            RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.UNEXPECTED_ERROR).build()
        }
    }

    fun handleAdd(guild: Guild, member: Member): MessageEmbed {
        val config = FavouriteTracksCache.instance
        val musicManager = RobertifyAudioManager[guild]
        val player = musicManager.player
        val playingTrack = player.playingTrack
        val memberVoiceState = member.voiceState!!
        val selfVoiceState = guild.selfMember.voiceState!!

        val acChecks = audioChannelChecks(memberVoiceState, selfVoiceState)
        if (acChecks != null) return acChecks

        val id = playingTrack.identifier

        logger.debug("source: ${playingTrack.source}")
        val source = when (playingTrack.source) {
            "spotify" -> TrackSource.SPOTIFY
            "deezer" -> TrackSource.DEEZER
            "applemusic" -> TrackSource.APPLE_MUSIC
            "soundcloud" -> TrackSource.SOUNDCLOUD
            "resume" -> TrackSource.RESUMED
            else -> run {
                return RobertifyEmbedUtils.embedMessage(
                    guild,
                    FavouriteTracksMessages.FT_INVALID_SOURCE
                ).build()
            }
        }

        return try {
            config.addTrack(
                uid = member.idLong,
                trackID = id,
                title = playingTrack.title,
                author = playingTrack.author,
                source = source
            )

            RobertifyEmbedUtils.embedMessage(
                guild,
                FavouriteTracksMessages.FAV_TRACK_ADDED,
                Pair("{title}", playingTrack.title),
                Pair("{author}", playingTrack.author)
            ).build()
        } catch (e: IllegalArgumentException) {
            RobertifyEmbedUtils.embedMessage(guild, e.message ?: "Invalid arguments!").build()
        } catch (e: Exception) {
            logger.error("Unexpected error", e)
            RobertifyEmbedUtils.embedMessage(guild, GeneralMessages.UNEXPECTED_ERROR).build()
        }
    }

    override suspend fun onStringSelectInteraction(event: StringSelectInteractionEvent) {
        if (event.selectMenu.id?.startsWith("menupage") == false) return
        if (event.user.id != event.componentId.split(":")[1]) return

        val selectionOption = event.selectedOptions.first().value
        if (!selectionOption.startsWith("favouriteTrack")) return

        val id = selectionOption.split(":")[1]
        val source = TrackSource.parse(selectionOption.split(":")[2])
        val audioManager = RobertifyAudioManager
        val guild = event.guild!!
        val memberVoiceState = event.member!!.voiceState!!
        val selfVoiceState = guild.selfMember.voiceState!!

        val acChecks = audioChannelChecks(memberVoiceState, selfVoiceState, selfChannelNeeded = false)
        if (acChecks != null) {
            event.replyEmbed { acChecks }
                .setEphemeral(true)
                .queue()
            return
        }

        event.replyEmbed {
            embed(FavouriteTracksMessages.FT_ADDING_TO_QUEUE)
        }.setEphemeral(true)
            .queue()

        val messageChannel = event.channel.asGuildMessageChannel()
        messageChannel.sendEmbed(guild) {
            embed(FavouriteTracksMessages.FT_ADDING_TO_QUEUE_2)
        }.queue { addingMsg ->
            val url = when (source) {
                TrackSource.DEEZER -> "https://www.deezer.com/us/track/$id"
                TrackSource.SPOTIFY -> "https://www.open.spotify.com/track/$id"
                TrackSource.APPLE_MUSIC -> "https://www.music.apple.com/us/song/$id"
                TrackSource.YOUTUBE,
                TrackSource.SOUNDCLOUD,
                TrackSource.RESUMED -> id
            }

            audioManager.loadAndPlay(
                trackUrl = url,
                memberVoiceState = memberVoiceState,
                botMessage = addingMsg
            )
        }
    }

    private fun setDefaultEmbed(
        member: Member,
        tracks: List<FavouriteTracksCache.Track>,
        theme: RobertifyTheme
    ) {
        val localeManager = LocaleManager.getLocaleManager(member.guild)
        PaginationHandler.embedStyle = {
            EmbedBuilder()
                .setColor(theme.color)
                .setTitle(
                    localeManager.getMessage(
                        FavouriteTracksMessages.FT_EMBED_TITLE,
                        Pair("{user}", member.effectiveName)
                    )
                )
                .setFooter(
                    localeManager.getMessage(
                        FavouriteTracksMessages.FT_EMBED_FOOTER,
                        Pair("{tracks}", tracks.size.toString())
                    )
                )
                .setDescription(localeManager.getMessage(FavouriteTracksMessages.FT_EMBED_DESCRIPTION))
                .setThumbnail(member.effectiveAvatarUrl)
        }
    }
}